[미션] CardUI-FETCHING

dev-angelo opened the pull request:
완료 사항 >

1. 필요한 형태의 Mock 데이터 구성 (JSON 형식). **(Mission)**
2. node.js express 로컬서버 환경 구성. **(Mission)**
3. 기존 SlideServie 에서 fetch API 를 활용해서 요청/응답 받도록 수정. **(Mission)**
4. 지난 animation 에서 리뷰 받았던 부분 수정 >
  -- 중재자에게 전달하던 message 의 상수화.
  -- SlideService 에서 컴포넌트 등록 시 SlideComponent 를 상속받은 객체인지 확인.

힘들었던 부분 >
1. 우선 기존 코드에 미션에 대한 기능을 완료 한 후 리팩토링을 해야 했으나
    처음부터 MVC 패턴에 대한 고민을 하다 코드를 작성하는 시간이 지체되었음.
2. 최대한 Component 를 독립적으로 동작하도록 만드려고 하였으나,
    render 하는 과정에서 생기는 dependency (html 에서의 block 속성 때문에
    생기는 의존성.. 해결 방법이 있는지 찾아보도록 하겠습니다.) 때문에
    의존성이 생겨버린 부분.

To do >
1. MVC 패턴에 대한 공부 및 다음 미션은 MVC 패턴으로 작성 해볼 수 있도록 연습.
2. Promise 패턴을 적용 할 수 있는 부분을 찾아 적용해보기.

dev-angelo committed:
[파일추가] .gitignore 파일 추가

.gitignore 파일 추가 >
* node module 들을 ingore 공식에 추가.

dev-angelo committed:
[파일추가] express 를 위한 의존성이 명시되어 있는 package 파일 추가

package-lock.json, package.json 파일 추가 >
* express 를 위한 의존성이 명시되어 있는 package 파일 추가.

dev-angelo committed:
[파일추가] MockData 를 response 해 주기 위한 파일 추가

Server.js >
* html 에 렌더링 하기 위해 필요한 localData.json 파일을 읽어들임.
* 요청이 왔을 때 JSON 으로 변환한 파일 애용을 response.

localData.json >
* Carousel 에 보여지기 위한 MockData 가 json 형식으로 명시된 파일.

dev-angelo committed:
[수정] Dom templating 이 용이하도록 수정

Dom templating 이 용이하도록 html, css, js 파일을 수정.

Why >
* 이전 미션에서는 데이터가 이미 존재한다는 상황이었으나
  이번 미션은 데이터를 요청하여 얻어온 다음
  dom templating 을 통해 렌더링을 해야 하기 때문.

* 이전 구조는 dom templating 을 하기 힘든 구조이기 때문에 수정 발생.

index.html >
* Dom templating 이 용이하도록 구조변경.

style.css >
* html 구조 변경이 되었지만 이전 스타일 유지되도록 수정.

main.js >
* html 구조 변경에 따라 content, menu button, direction 버튼 들의
  객체를 얻어오기 위한 동작 수정.

dev-angelo committed:
[수정] Server 에서 CORS 를 허용하도록 수정

Server 에서 CORS 를 허용하도록 수정.

Why >
* 포트가 다른 서버의 자원을 요청하여 이슈가 발생.

package.json, package-lock.json >
* CORS 관련 정보가 추가되었음.

Server.js >
* CORS 를 허가하는 코드 추가.

dev-angelo committed:
[수정] 서버로의 데이터 요청 요청 & 응답에 대한 동작이 가능해지도록 수정

Why >
* 기존에는 데이터가 있다는것을 전제로 Slide 가 설계되어 있었음.

SlideService.js >
1. 서버에 데이터를 request 하는 코드 추가.
2. 서버의 response 를 받았을 때 이를 각 컴포넌트로 전달하는 코드 추가.
3. 2 의 동작이 끝난 후 각 컴포넌트의 render() 를 호출하여
   Dom Templating 을 수행하는 코드 추가.
4. 컴포넌트 객체 생성 시점을 SlideService 생성자로 이동.

SlideComponent.js >
1. 생성자로 전달되었던 'elements' 매개변수 삭제.
-> DOM Templating 이전에는 elements 에 대한 내용을 얻어올 수 없기 때문.

2. _registerEventListenerOnElements() 를 생성자에서 호출하지 않도록
   수정.
-> 기존에는 데이터가 있다고 가정하고 바로 DOM handling 을 통해
   객체를 얻어올 수 있었지만, 이제는 Dom Templating 이후
   객체를 얻어와야 하기 때문.)

3. onNotifyDataChanged(), onNotifyRenderFinished(), render() 인터페이스
   추가.
-> onNotifyDataChanged(): 서버에서 데이터를 받아 온 후, 데이터가
   바뀌었다는 사실을 중재자가 Component 들에게 알리기 위한 인터페이스.

-> render(): 서버로부터 바뀐 데이터를 컴포넌트들에게 전달 한 후
   Dom Templating 을 위해 각 컴포넌트들의 Dom Templating 을
   전달받기 위한 인터페이스.

-> onNotifyRenderFinished(): 렌더 완료 후 Component 들에게 그 사실을
   알리기 위한 인터페이스.

MenuButtonManager.js, DirectionButtonManager.js >
1. 생성자로 전달되었던 'elements' 매개변수 삭제.
-> SlideComponent 에서 삭제 되었기 때문.

2. onNotifyDataChanged(), onNotifyRenderFinished(), render()
   함수 재정의 추가
-> onNotifyDataChanged(): 함수가 불릴 때 전달된 data 를
   내부적으로 갖고 있도록 수정.
-> onNotifyRenderFinished(): 중재자로부터 render 가 완료 되었다는
   사실을 전달 받았을 때 각 manager 에서 다룰 button 객체를
   얻어오고, 객체에 event listener 를 추가하는 코드 작성.
-> render(): onNotifyDataChanged() 를 통해 얻어온 정보를 토대로
   Dom Templating 을 위한 string 작성.

main.js >
1. 동작방식 변경에 의해 필요 없어진 코드들 삭제.

index.html >
1. 동작방식 변경에 의해 필요 없어진 코드들 삭제.

localData.json >
1. Menu 에 대한 정보 추가.

dev-angelo committed:
[삭제] mock data 관련 파일 삭제

mock data 관련 파일 삭제.

Why >
1. 서버에 요청을 하여 데이터를 받아오기 때문.

dev-angelo committed:
[수정] 컴포넌트 등록 시 유효성 확인

컴포넌트 등록 시 SlideComponent 를 상속받은
객체인지 확인하는 코드 추가.

Why >
SlideComponent 를 상속받은 객체가 아닐 때 등록이 되면
SlideComponent 의 interface 가 재정의되어 있지 않기 때문에
정상적인 동작을 하지 않기 때문.

dev-angelo committed:
[추가] SlideEnum.js 파일 추가

Slide 동작을 상수화 하기 위한 Enum 추가.

dev-angelo committed:
[수정] SlideEnum 을 사용하도록 수정

SlideEnum 을 사용하도록 수정.

Why >
상수화되지 않은 message 를 주고 받을 경우
human 에러가 나기 쉽고 변경에도 유연하지 않음.

dev-angelo committed:
[수정] 불필요한 코드 삭제

불필요한 코드 삭제.

dev-angelo committed:
[수정] mediate() 메서드의 매개변수 이름 수정

mediate() 메서드의 매개변수인 message 의 네이밍을
target 으로 변경.

dev-angelo committed:
[수정] SlideEnum 수정, localStorage 캐싱 되도록 수정

SlideEnum 수정 >
* 잘못된 SlideEnum 작성으로 인해 버튼들 오동작 하는 문제 수정.

localStorage 에 캐싱 되도록 수정 >
* localStroage 에 캐싱 된 데이터를 재 사용하지 않았던 문제 수정.

dev-angelo committed:
[수정] 불필요한 console.log() 삭제

불필요한 console.log() 삭제

dev-angelo committed:
[수정] 초기화 함수에서 templating 기능 분리

_initialize() 함수에서 templating 기능을 하는 코드를
_template() 함수로 이동.

dev-angelo committed:
[수정] localStorage 캐싱 관련 동작 수정

localStorage 캐싱 관련 동작 수정

SlideService.js >
1. slideData 가 캐시되어 있지 않을 때 (최초접속)
   json data 를 정상적으로 처리하지 못했던 문제 수정.

DongHyunKims commented on MenuButtonManager.js:
```diff
@@ -26,6 +27,26 @@ class MenuButtonManager extends SlideComponent {
+    }
+
+    render() {
+        let result = `<ul id="menu">`;
```
tagged template literal을 사용할 필요가 없다면 일반 string으로 작성하는게 좋아보여요!

DongHyunKims commented on MenuButtonManager.js:
```diff
@@ -26,6 +27,26 @@ class MenuButtonManager extends SlideComponent {
+            result += `<li><button id="button${i}">${this._data[i].title}</button></li>`;
+        }
+
+        result += `</ul>`
```
여기두요~

DongHyunKims commented on MenuButtonManager.js:
```diff
@@ -17,7 +18,7 @@ class MenuButtonManager extends SlideComponent {
     }
 
     _menuButtonHandler(event, slideService, index) {
```
어디서 사용하는 메소드 일까요? 사용하는 곳이 없어보이는데요~

dev-angelo replied:
_menuButtonHandler() 및 _directionButtonHandler() 는_registerEventListenerOnElement() 메소드가 호출 될 때 메소드 내부에서 호출되는 메소드였습니다.

아래 리뷰 (상속 받는 곳에서 사용하는 모든 동작이 다른가요?~) 를 통해 수정된 코드의 대략적인 모습은 아래와 같습니다.

```
SlideComponent.js (Before)
    _registerEventListenerOnElements(elements) {
    }
```

```
SlideComponent.js (After)
    _registerEventListenerOnElements(elements) {
        elements.forEach((element, index) => {
            element.addEventListener('click', event => {
                this._appendButtonHandler(event, this._slideService, index);
            });
        });
    }
```

```
MenuButtonManager.js, DirectionButtonManager.js (Before)

    (아래 메소드 내용이 위처럼 SlideComponent.js 로 이동)
    _registerEventListenerOnElements(slideService, elements) {
        elements.forEach((element, index) => {
            element.addEventListener('click', event => {
                this._directionButtonHandler(event, slideService, index);
            });
        });
    }

   _menuButtonHandler(..) (_directionButtonHandler) {
       ~~
   }
```

```
MenuButtonManager.js, DirectionButtonManager.js (After)

    //_registerEventListenerOnElements(slideService, elements) {
    //    elements.forEach((element, index) => {
    //        element.addEventListener('click', event => {
    //            this._directionButtonHandler(event, slideService, index);
    //        });
    //    });
    //}

   _appendButtonHandler (..) {
       ~~
   }
```

DongHyunKims commented on SlideComponent.js:
```diff
@@ -1,14 +1,23 @@
+    }
+
     _registerEventListenerOnElements(elements) {
     }
```
상속 받는 곳에서 사용하는 모든 동작이 다른가요? 최대한 공통으로 사용할수 있는 메소드는 작성해 주는 것이 좋지 않을까요?

dev-angelo replied:
슬라이드를 flicking 동작을 이용하여 움직이는 요구사항이 있어야 한다면 되면 상속는 곳에서 동작이 달라지게 되는데, 현재는 버튼 click 만을 위한 동작만 하게 되어서 동작 정의를 SlideComponent 에 올리도록 하겠습니다~!

DongHyunKims commented on SlideService.js:
```diff
@@ -1,37 +1,98 @@
+            fetch(url)
+                .then(response => response.json())
+                .then(slideData => {
+                    this._setLocalstorageData(JSON.stringify(slideData));
+                    this._initialize(JSON.stringify(slideData));
```
모든 데이터 리스트를 local storage에 저장 해놓을 필요가 있을지 고민해 봤으면 좋겠어요~ 어차피 로드전 한번만 fetch할것 같아서요~ 
현재 index와 같은 간단한 데이터만 저장해놓고 사용하는건 어떨까요?

dev-angelo replied:
의견 주신 부분이 잘 이해가 가지 않아서 질문 드립니다 ㅎㅎ;
index 저장이 현재 화면에 나오고 있는 내용의 current index 를 캐싱 의미하는것인가요?

crongro replied:
네 current index를 말씀하신거 같네요 ~

DongHyunKims commented on SlideService.js:
```diff
@@ -1,37 +1,98 @@
-    mediate(message, index) {
-        if ('increaseCurrentIndex' === message) {
+    mediate(target, index) {
+        if (SlideEnum.INCREASE_CURRENT_INDEX === target) {
```
변수명만 `target`으로 변경했는데 가독성이 굉장히 좋아졌네요~ 👍

DongHyunKims commented on SlideComponent.js:
```diff
@@ -1,14 +1,23 @@
 class SlideComponent {
```
인터페이스를 생각하고 만들었다면 상속하지 않았을시 에러가 제공되었으면 좋겠네요~

dev-angelo replied:
SlideService.js 파일 안에서 SlideComponent 를 상속받지 않은 컴포넌트가 등록되려고 할 때에 대한 처리를 했는데 에러 처리하는게 더 명시적일것 같습니다.

```
registerComponent(component) {
    if (!component instanceof SlideComponent)
        return;

    this._components.push(component);
}
```

단순 리턴 하는것이 아닌 에러를 반환하도록 수정해보겠습니다!

DongHyunKims reviewed (commented):
한주동안 수고하셨어요~ 컴포넌트를 나누고 상속화해서 사용하려는 노력이 보기 좋아요~

dev-angelo committed:
[수정] button event handler 등록 관련 수정

button event handler 등록 관련 수정.

SlideComponent.js >
1. MenuButtonManager 및 DirectionButtonManager 의
   _registerEventListenerOnElements() 동작이 같으므로
   _registerEventListenerOnElements() 코드를
   SlideComponent 로 이동

2. SlideComponent 를 상속받는 Component 들에서 재 정의 될
   _appendButtonHandler() 인터페이스 추가.

MenuButtonManager.js, DirectionButtonManager.js >
1. _registerEventListenerOnElements() 동작이 상위 class 로
   이동했으므로 재정의 하던 부분 삭제.

2. button 에 click 발생 시 호출되는 함수를 _appendButtonHandler() 로
   수정.

crongro commented:
머지처리했어요~
@dev-angelo

dev-angelo commented:
> 머지처리했어요~
> @dev-angelo

감사합니다~!
